File: python/customer.py
from price import MovieCode
from rental import Rental
from movie import Movie


class Customer:

    def __init__(self, name):
        self._name = name
        self._rentals = []

    def add_rental(self, arg):
        self._rentals.append(arg)

    @property
    def name(self):
        return self._name

    @property
    def rentals(self):
        return self._rentals

    def statement(self):
        result = 'Rental record for ' + self.name + '\n'

        for rental in self.rentals:
            result += '\t' + rental.movie.title + '\t' + str(rental.cost()) + '\n'

        result += 'You owed ' + str(self.calculate_total_cost()) + '\n' \
                  + 'You earned ' + str(self.calculate_total_point()) + ' points\n'
        return result

    def html_statement(self):
        result = '<h1><em>Rental record for ' + self.name + '</em></h1><p>\n'

        for rental in self.rentals:
            result += rental.movie.title + ': ' + str(rental.cost()) + '<br>\n'

        result += '<p>You owed <em>' + str(self.calculate_total_cost()) + '</em><p>\n' \
                  + 'You earned <em>' + str(self.calculate_total_point()) + '</em> points<p>\n'
        return result

    def calculate_total_cost(self):
        result = 0
        for rental in self.rentals:
            result += rental.cost()
        return result

    def calculate_total_point(self):
        result = 0
        for rental in self.rentals:
            result += rental.points()
        return result


if __name__ == '__main__':
    customer = Customer("Jamie")
    movies = [Movie("Interstella", MovieCode.REGULAR),
              Movie("Arrival", MovieCode.NEW_RELEASE),
              Movie("Moana", MovieCode.CHILDREN),
              Movie("LaLaLand", MovieCode.NEW_RELEASE)]

    for i in range(4):
        customer.add_rental(Rental(movies[i], i+3))

    print(customer.statement())

File: java/src/main/videostore/Customer.java
package videostore;

import java.util.Enumeration;
import java.util.Vector;

class Customer {
    private String _name;
    private Vector _rentals = new Vector();

    public Customer(String name) {
        _name = name;
    }

    public void addRental(Rental arg) {
        _rentals.addElement(arg);
    }

    public String getName() {
        return _name;
    }

    public String statement() {
        Enumeration rentals = _rentals.elements();
        String result = "Rental Record for " + getName() + "\n";
        while (rentals.hasMoreElements()) {
            Rental each = (Rental) rentals.nextElement();

            result += "\t" + each.getMovie().getTitle() + "\t" + String.valueOf(each.getCharge()) + "\n";
        }

        result += "You owed " + String.valueOf(getTotalCharge()) + "\n";
        result += "You earned " + String.valueOf(getTotalFrequentRenterPoints()) + " frequent renter points\n";

        return result;
    }

    public String htmlStatement() {
        Enumeration rentals = _rentals.elements();
        String result = "<h1><em>Rental Record for " + getName() + "</em><h1><p>\n";
        while (rentals.hasMoreElements()) {
            Rental each = (Rental) rentals.nextElement();

            result += "\t" + each.getMovie().getTitle() + ": " + String.valueOf(each.getCharge()) + "<br>\n";
        }

        result += "<p> You owed <em>" + String.valueOf(getTotalCharge()) + "</em><p>\n";
        result += "You earned <em>" + String.valueOf(getTotalFrequentRenterPoints()) + "</em> frequent renter points<p>\n";
        return result;
    }

    private double getTotalCharge() {
        double result = 0;
        Enumeration rentals = _rentals.elements();
        while (rentals.hasMoreElements()) {
            Rental each = (Rental) rentals.nextElement();
            result += each.getCharge();
        }
        return result;
    }

    private int getTotalFrequentRenterPoints() {
        int result = 0;
        Enumeration rentals = _rentals.elements();
        while (rentals.hasMoreElements()) {
            Rental each = (Rental) rentals.nextElement();
            result += each.getFrequentRenterPoints();
        }
        return result;
    }
}